prow/cmd/deck/main.go

```diff
@@ -1172,29 +1202,33 @@ func handleRerun(prowJobClient prowv1.ProwJobInterface, createProwJob bool) http
 		// On Prow instances that require auth even for viewing Deck this is okayish, because the Prowjob UUID
 		// is hard to guess
 		// Ref: https://github.com/kubernetes/test-infra/pull/12827#issuecomment-502850414
-		if createProwJob {
-			if r.Method != http.MethodPost {
-				http.Error(w, "request must be of type POST", http.StatusMethodNotAllowed)
+		if r.Method == http.MethodPost {
+			if !createProwJob {
+				http.Error(w, "Direct rerun feature is not yet enabled", http.StatusMethodNotAllowed)
+				return
+			}
+			login, err := r.Cookie("github_login")
+			if err == http.ErrNoCookie {
+				http.Redirect(w, r, "/github-login?dest=%2F?rerun=work_in_progress", http.StatusFound)
+				return
+			} else if err != nil {
+				http.Error(w, fmt.Sprintf("Error retrieving GitHub cookie: %v", err), http.StatusInternalServerError)
+				logrus.WithError(err).Errorf("Error retrieving GitHub cookie")
+				return
+			}
+			if !canTriggerJob(login.Value, cfg().Deck.RerunAuthConfig) {
+				http.Redirect(w, r, "/?rerun=error", http.StatusFound)
 				return
 			}
 			if _, err := prowJobClient.Create(&newPJ); err != nil {
 				logrus.WithError(err).Error("Error creating job")
 				http.Error(w, fmt.Sprintf("Error creating job: %v", err), http.StatusInternalServerError)
 				return
 			}
-			w.WriteHeader(http.StatusNoContent)
+			http.Redirect(w, r, "/?rerun=success", http.StatusFound)
 			return
 		}
-		b, err := yaml.Marshal(&newPJ)
-		if err != nil {
-			http.Error(w, fmt.Sprintf("Error marshaling: %v", err), http.StatusInternalServerError)
-			logrus.WithError(err).Error("Error marshaling jobs.")
-			return
-		}
-		w.Header().Set("Content-Type", "application/json")
-		if _, err := w.Write(b); err != nil {
-			logrus.WithError(err).Error("Error writing log.")
-		}
+		marshalJob(w, newPJ)
```

Comment:
This looks better but still isn't following the principle of returning early and minimizing indentation compared to something like this:

```golang

switch r.Method {
case http.MethodGet:
  return marshalJob(w, newPJ)
case http.MethodPost:
  // acceptable
default:
  return fmt.Errorf("bad verb %v", r.Method")
}

// logic for post
```